Source/Core/Core/HW/DVD/AMMediaboard.cpp

```diff
@@ -0,0 +1,1909 @@
+#include <string>
+#include <unordered_map>
+
+#include <fmt/format.h>
+
+#include "Common/Buffer.h"
+#include "Common/CommonTypes.h"
+#include "Common/FileUtil.h"
+#include "Common/IOFile.h"
+#include "Common/Logging/Log.h"
+
+#include "Core/Boot/Boot.h"
+#include "Core/BootManager.h"
+#include "Core/ConfigManager.h"
+#include "Core/Core.h"
+#include "Core/HLE/HLE.h"
+#include "Core/HW/DVD/DVDInterface.h"
+#include "Core/HW/DVD/DVDThread.h"
+#include "Core/HW/EXI/EXI.h"
+#include "Core/HW/EXI/EXI_DeviceBaseboard.h"
+#include "Core/HW/MMIO.h"
+#include "Core/HW/Memmap.h"
+#include "Core/HW/SI/SI.h"
+#include "Core/HW/SI/SI_Device.h"
+#include "Core/IOS/Network/Socket.h"
+#include "Core/Movie.h"
+#include "Core/System.h"
+
+#include "DiscIO/DirectoryBlob.h"
+
+#if defined(__linux__) or defined(__APPLE__) or defined(__FreeBSD__) or defined(__NetBSD__) or     \
+    defined(__HAIKU__)
+
+#include <unistd.h>
+
+static constexpr auto* closesocket = close;
+static auto ioctlsocket(auto... args)
+{
+  return ioctl(args...);
+}
+
+static constexpr int WSAEWOULDBLOCK = 10035;
+static constexpr int SOCKET_ERROR = -1;
+
+using SOCKET = int;
+
+static constexpr SOCKET INVALID_SOCKET = SOCKET(~0);
+
+static int WSAGetLastError()
+{
+  switch (errno)
+  {
+  case EINPROGRESS:
+  case EWOULDBLOCK:
+    return WSAEWOULDBLOCK;
+  default:
+    break;
+  }
+
+  return errno;
+}
+
+#endif
+
+namespace AMMediaboard
+{
+
+static bool s_firmware_map = false;
+static bool s_test_menu = false;
+static SOCKET s_fd_namco_cam = 0;
+static std::array<u32, 3> s_timeouts = {20000, 20000, 20000};
+static u32 s_last_error = SSC_SUCCESS;
+
+static u32 s_gcam_key_a = 0;
+static u32 s_gcam_key_b = 0;
+static u32 s_gcam_key_c = 0;
+
+static File::IOFile s_netcfg;
+static File::IOFile s_netctrl;
+static File::IOFile s_extra;
+static File::IOFile s_backup;
+static File::IOFile s_dimm;
+
+static Common::UniqueBuffer<u8> s_dimm_disc;
+
+auto& s_system = Core::System::GetInstance();
+auto& s_memory = s_system.GetMemory();
+
+static u8 s_firmware[2 * 1024 * 1024];
+static u32 s_media_buffer_32[192];
+static u8* const s_media_buffer = reinterpret_cast<u8*>(s_media_buffer_32);
+static u8 s_network_command_buffer[0x4FFE00];
+static u8 s_network_buffer[512 * 1024];
+static u8 s_allnet_buffer[4096];
+static u8 s_allnet_settings[0x8500];
+
+constexpr char s_allnet_reply[] = {
+    "uri=http://"
+    "sega.com&host=sega.com&nickname=sega&name=sega&year=2025&month=08&day=16&hour=21&minute=10&"
+    "second=12&place_id=1234&setting=0x123&region0=jap&region_name0=japan&region_name1=usa&region_"
+    "name2=asia&region_name3=export&end"};
+
+// Sockets FDs are required to go from 0 to 63.
+// Games use the FD as indexes so we have to workaround it.
+
+static SOCKET s_sockets[64];
+
+static u32 SocketCheck(u32 x)
+{
+  if (x < std::size(s_sockets))
+    return x;
+
+  WARN_LOG_FMT(AMMEDIABOARD, "GC-AM: Bad SOCKET value: {}", x);
+  return 0;
+}
+
+static bool NetworkCMDBufferCheck(u32 x)
```

Comment:
This needs to also have a length parameter so that you can check that offset + length is in bounds.